provider/core/src/resource.rs

```diff
@@ -74,59 +73,132 @@ impl writeable::Writeable for ResourceCategory {
+/// Executables can be searched for ResourceKey instances to produce optimized data files.
+/// Therefore, users should not generally create ResourceKey instances; they should instead use
+/// the ones exported by a component.
+#[derive(PartialEq, Eq, Copy, Clone)]
+#[repr(C)]
 pub struct ResourceKey {
-    pub category: ResourceCategory,
-    pub sub_category: TinyStr16,
-    pub version: u16,
+    path: &'static str,
+    _tag0: [u8; 8],
+    hash: ResourceKeyHash,
+    _tag1: [u8; 2],
 }
 
-/// Shortcut to construct a const resource identifier.
-///
-/// # Examples
-///
-/// Create a private-use ResourceKey:
-///
-/// ```
-/// use icu_provider::prelude::*;
-///
-/// const MY_PRIVATE_USE_KEY: ResourceKey = icu_provider::resource_key!(x, "foo", "bar", 1);
-/// assert_eq!("x-foo/bar@1", format!("{}", MY_PRIVATE_USE_KEY));
-/// ```
-///
-/// Create a ResourceKey for a specific [`ResourceCategory`] (for ICU4X library code only):
-///
-/// ```
-/// use icu_provider::prelude::*;
-///
-/// const MY_PRIVATE_USE_KEY: ResourceKey = icu_provider::resource_key!(Plurals, "ordinal", 1);
-/// assert_eq!("plurals/ordinal@1", format!("{}", MY_PRIVATE_USE_KEY));
-/// ```
-#[macro_export]
-macro_rules! resource_key {
-    ($category:ident, $sub_category:literal, $version:tt) => {
-        $crate::resource_key!($crate::ResourceCategory::$category, $sub_category, $version)
-    };
-    (x, $pu:literal, $sub_category:literal, $version:tt) => {
-        $crate::resource_key!(
-            $crate::ResourceCategory::PrivateUse($crate::internal::tinystr4!($pu)),
-            $sub_category,
-            $version
-        )
-    };
-    ($category:expr, $sub_category:literal, $version:tt) => {
-        $crate::ResourceKey {
-            category: $category,
-            sub_category: $crate::internal::tinystr16!($sub_category),
-            version: $version,
+impl ResourceKey {
+    /// Gets a human-readable representation of a [`ResourceKey`].
+    /// 
+    /// The human-readable path string always contains at least one '/', and it ends with '@'
+    /// followed by one or more digits. Paths do not contain characters other than lowercase
+    /// ASCII, '_', '/', and '@'.
+    /// 
+    /// Useful for reading and writing data to a file system.
+    #[inline]
+    pub const fn get_path(&self) -> &str {
+        self.path
+    }
+
+    /// Gets a machine-readable representation of a [`ResourceKey`].
+    ///
+    /// The machine-readable hash is 4 bytes and can be used as the key in a map.
+    ///
+    /// The hash is a 32-bit FxHash of the path, computed as if on a little-endian platform.
+    #[inline]
+    pub const fn get_hash(&self) -> ResourceKeyHash {
+        self.hash
+    }
+
+    /// Creates a new ResourceKey from a path, panicking if the path is invalid.
+    /// 
+    /// # Panics
+    /// 
+    /// Panics if the syntax of the path is not valid.
+    #[inline]
+    pub const fn new(path: &'static str) -> Self {
```

Comment:
I did it this way because:

1. This new model allows multiple layers of key strings, rather than just category/subcategory, which was an artificial limitation
2. I find that writing the string in code is actually shorter and more readable than using an opaque concatenation algorithm